Changes for transfer IDs logic

vkuznet opened the pull request:
Partially fixes #9196 

#### Status
In development

#### Description
It implements logic for MS based on transfer IDs. Here is how it is done:
  - code refactoring
     - `MSCore` is a base class for `MSMonitor`, `MSTransferor` and `RequestInfo`
  - transferor API obtain request info and perform transferRequest
    the transferRequest performs phedex subscription (via subscribe API which
    so far is commented out) and obtain transfer ids via getTransferIds API
    It does this for every dataset in request spec. At the end transfer API
    received status (transfer) records which its store to couchdb via
    `updateTransferInfo` API (which calls apis from 9275)
  - monit API get requests and fetch status (transfer) records from couchdb
    via `getTransferRecords` API (which by itself calls getTransferIds one).
    Then we check completion of transfer via `completion` API. If completed value is
    100 (percent) we are done, otherwise we update new set of records
    in couchdb via updateTransferInfo API.

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
depends on #9198 , #9275, #9307 

#### External dependencies / deployment changes
no

vkuznet committed:
Changes for transfer IDs logic

amaltaro commented:
As mentioned in the meeting, please create an issue reporting the problem and what's the expected behavior.

vkuznet commented:
Alan, this PR already listed dependencies:  #9198 , #9275. The only missing bit is you need to decide which DB name to use and which structure (transfer IDs or suscription) we'll deal with. Once settle and we both issues will be merged I can move with this.

amaltaro commented:
Valentin, I'm having a hard time reviewing this PR. Can you please describe in the first comment box what exactly are these changes? What is the flow of actions executed by MS here? and what has to be discussed/defined/fix ahead.

vkuznet commented:
let's adjust it once you'll clarify which Phedex APIs to use.

amaltaro commented on src/python/WMCore/MicroService/Unified/RequestInfo.py:
```diff
@@ -461,8 +461,10 @@ def unified(svc, requestRecords, logger):
-            ndatasets = len(attrs['datasets'])
-            npileups = len(attrs['pileups'])
+            campaign = attrs['campaign']
+            campaignType = 'primary' # TODO: how to define it?
```
That's a good question. I believe it's defined in this method:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/Unified/MSTransferor.py#L115

when it's looking for all possible data that we need to transfer (it misses the parent though). So we should not hardcode it but actually see all possible values and get it.

amaltaro commented on src/python/WMCore/MicroService/Unified/Transferor.py:
```diff
@@ -143,79 +167,147 @@ def stop(self):
-    def requestCampaign(self, req):
-        "Return request campaign"
-        return 'campaign_TODO' # TODO
+    def getStatusRecords(self, reqName):
```
Hmm, we don't really persist a status of the transfers. It will be calculated on the fly based on the transfer document. OR, we need to add another key to the trasnfer document schema to keep track of the completion status of those transfers.

Then we still have to define: "completion", but that's in another time.

amaltaro reviewed (changes requested):
Valentin, I've done just an initial review because this PR needs to be rebased (I pushed in a bunch of changes on the last Friday).

Nonetheless, let me list the steps we should take in the monitoring thread, in addition to what's described in the point 6. of this wiki:
https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Transferor#relying-on-reqmgr2-second-proposal

The step by step would be:
1. fetch all the workflows in the `staging` status
2. fetch all the workflow transfer documents
2.a. filter those documents out such that we don't check for their completion status every single cycle. So we could do something like `if timestamp - time() > X hours: then check this one`. I'd say every 4h or so.
3. fetch the campaign configuration
4. using the transfer ids, make calls to phedex and get the transfer status
5. calculate the status completion
5.a. if ALL transfer type (primary, secondary, etc) are above the minimum thresholds, then a) update the transfer doc and b) update the request status to `staged`
5.b. if NOT ALL transfers are above the minimum thresholds, then update the transfer doc and proceed to the next workflow
5.c. IF there are no transfers under the workflow transfer doc (it means there is not any type of input data), then move the workflow to `staged`

I'll update the wiki page soon. Please let me know if you have any questions.

vkuznet commented:
ok, then I'll rebase once you'll merge https://github.com/dmwm/WMCore/pull/9307 and continue working on this

amaltaro commented:
Done, feel free to proceed with this one. I'm going to ask a new testbed release tomorrow morning, so we might have to postpone it until the next week (if safe changes).

vkuznet commented:
Alan, do you mind to explain why did you re-factor MS code and what exactly you did.
I see SO MANY changes that I want to understand the entire logic before proceeding. For instance, to preserve previous work I need clear understanding which branch to use to resolve the conflicts, the head or the one I had before. I need to understand which methods/functions to keep, etc.

And, for the future, please elaborate and discuss the changes up-front. I already rewrote this piece so many times that I start having a feeling that I'm wasting my time on this codebase.

amaltaro commented:
Because the transferor code was looking an all-in-one, where we were piling up methods and more methods without really thinking where things belong to, without reusing code.

I have all that explained in the first comment of this PR:
https://github.com/dmwm/WMCore/pull/9296

where you were tagged. I also asked you to have a closer look once you get back from vacation (it was merged in the past week). Let me know if you still have questions, but we better then follow up on that PR to keep the discussion in the same context.

amaltaro commented:
I have started the basic description/behaviour of the Monitor thread here:
https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Monitor

but as you know, we might have to touch it once we fully understand the "black box" logic of the transferor service and all the parameters it depends on. Which made me create this option D here:
https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154#file-transferor_format-json-L32

and add an extra field `campaignName`, such that we don;t need to look into a request detail once again in the monitor thread.

vkuznet commented:
I performed rebase and added/fixed the following:
- I re-factored all mine and Alan's changes to separate MS logic into the following classes:
  - MSCore.py represents core module for MSTransferor, MSMonitor, RequestInfo classes
  - I separated transferor, monitoring and request info into separate classes
  - I cleaned and implemented consistent logic in MSManager class
- I refactored Alan's changes to make code thread safe (especially introduced in RequestInfo and MSManager classes)
- I implemented logic based on option D from transfer records document (see https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154)

I did not (yet) made any effort to complete full logic of transferor/monitoring MS services and suggest that it should be handled in next PR once we agree on these changes and merge them. I also didn't touch yet unified logic and suggest that it should be done after full MS logic will be in place.

amaltaro commented:
@vkuznet you probably have a broken import here, that's why this avalanche of unit tests failing.

In addition to that, please do not disable pylint checks blindly, otherwise we never know what needs attention in the future or not.

vkuznet commented:
Alan, I doubt it is broken import since pylint/pycodestyle will capture this. I also checked manually imports and they work fine. Plus I have doubt that such unit test:
```
WMCore_t.Services_t.DBS_t.DBSReader_t.DBSReaderTest:testListDatatiers changed from success to error
```
will be related to MS code. I think failure of unit tests is related by something else. Do you have any log to look what is exactly fails?

Regarding pylint checks, I don't exclude blindly. I excluded those which I believe irrelevant. For instance, here is a typical codes:

```
pylint --help-msg=C0103,R0914,R0912,R1702,R0201,W0703
No config file found, using default configuration
:invalid-name (C0103): *%s name "%s" doesn't conform to %s*
  Used when the name doesn't conform to naming rules associated to its type
  (constant, variable, class...). This message belongs to the basic checker.

:too-many-locals (R0914): *Too many local variables (%s/%s)*
  Used when a function or method has too many local variables. This message
  belongs to the design checker.

:too-many-branches (R0912): *Too many branches (%s/%s)*
  Used when a function or method has too many branches, making it hard to
  follow. This message belongs to the design checker.

:too-many-nested-blocks (R1702): *Too many nested blocks (%s/%s)*
  Used when a function or a method has too many nested blocks. This makes the
  code less understandable and maintainable. This message belongs to the
  refactoring checker.

:no-self-use (R0201): *Method could be a function*
  Used when a method doesn't use its bound instance, and so could be written as
  a function. This message belongs to the classes checker.

:broad-except (W0703): *Catching too general exception %s*
  Used when an except catches a too general exception, possibly burying
  unrelated errors. This message belongs to the exceptions checker.
```
If you want I can remove all pylints, but it is trivial to check, as shown, what codes means.

amaltaro commented:
Besides C0103, I'd say all the others have a more meaningful impact on the code and can eventually be made better (even C0103 could, after all most of it is new code...). If you disable them, we never get it reported (unless we remember to check the pylints that were disabled).
I'm just saying we don't need to have a perfect score, but if possible not to make it worse...

About the error, if you follow the `default` `Details` link, it brings you to a summary of all the unit tests failing, e.g.:
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/9231/testReport/junit/WMCore_t.WorkQueue_t.Policy_t.Start_t.Block_t/BlockTestCase/testLumiMask/

so the problem is not an import indeed, but the PhEDex module being mocked at the wrong place (you reshuffled things, so we don't need to mock PhEDEx there anymore). Ping me on slack if you need help walking through those reports.

vkuznet commented:
Alan, I removed pylint exceptions, I fixed unit test. Please review since I don't to waste time fixing either pylint or irrelevant unit tests (Rucio).

amaltaro commented:
Has it been tested? I need to request another WMCore tag tomorrow early in the morning and probably won't manage to have a thorough review and test.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSCore.py:
```diff
@@ -0,0 +1,52 @@
+    MSTransferor and MSMonitor classes.
+    """
+
+    def __init__(self, msConfig=None, logger=None):
```
make `msConfig` mandatory and remove the check below

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSCore.py:
```diff
@@ -0,0 +1,52 @@
+        Provides setup for MSTransferor and MSMonitor classes
+
+        :param config: MS service configuration
+        :param logger:
```
either add a description or remove this line :)

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSManager.py:
```diff
@@ -1,21 +1,34 @@
 Author     : Valentin Kuznetsov <vkuznet AT gmail dot com>
-Description: UnifiedTransferorManager class provides full functionality of the UnifiedTransferor service.
+Description: MSManager class provides full functionality of the MSManager service.
+It provides a REST interface to reqmgr2ms service and should be
```
The REST interface is actually provided under the Services package, no?

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -1,102 +1,83 @@
-
-Class to hold the whole logic behind the transferor module
+File       : MSTransferor.py
+Author     : Valentin Kuznetsov <vkuznet AT gmail dot com>
```
How come you put your name on the module I created? :)

vkuznet replied:
hmm, I thought I created MS codebase, you just mess it up :) But since you asked I put your name and will blame you for not working code from now on.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -1,102 +1,83 @@
         """
         Executes the whole transferor logic
-        :param reqStatus: request status to that matters for this module
+        :param reqStatus: request statue to process
```
typo

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -1,102 +1,83 @@
+            for dataset, tids in sdict.items():
+                statusRecord = {'timestamp': time.time(), 'dataset': dataset,
+                                'dataType': ctype, 'transferIDs': tids}
+                requestStatuses.setdefaul(wname, []).append(statusRecord)
```
typo

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -1,102 +1,83 @@
+                len(requestRecords), reqStatus)
+            # get complete requests information
+            # based on Unified Transferor logic
+            requestRecords = self.reqInfo(requestRecords)
```
You have just reverted the slicing changes I made. 
This code can potentially blow up memory usage on the node, since it collects tons of information for each request and only in the very end it would perform the data transfers. Please put the slicing logic back here.

vkuznet replied:
Alan, I was trying to merge broken branch and clean-up the conflicts. Please explicitly point out which slicing you're talking about and we can put it back. Better, I suggest to make another PR once this one will be merged since we first need to get proper codebase without conflicts.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -126,28 +107,24 @@ def requestRecord(self, wfData):
+                "### creating subscription for: %s", pformat(req))
+            subscription = PhEDExSubscription(
+                datasets, sites, self.msConfig['group'])
+            self.logger.info(
```
Breaking all these lines make the code harder to read IMO. In general, up to 80 chars in the same line is fine, if it goes beyond that, we might either break into multiple lines or not, it depends on how readable it is.

vkuznet replied:
please be concrete what you want, the code goes above 80 characters, that's why I break it according to pylint suggestion. So either comply with pylint or not but don't ask for both :)

amaltaro reviewed (changes requested):
I've performed a spot review and the code looks much better now!
There are a few issues that need to be solved though.

vkuznet commented:
Alan, I changed code according to your suggestions, but I didn't put slicing. Please provide a pointer and I can added it back. But once again, don't try to get ideal code until we understand its flow. Don't over optimize it until it is working, etc. I DO UNDERSTAND issue with memory consumption but we're still battling on codebase flow.

For the record code is not tested since I'm awaiting your ok with its logic.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -1,102 +1,84 @@
 
-        try:
-            reqInfo = RequestInfo(self.msConfig, self.uConfig, self.logger)
-            for reqSlice in grouper(requestRecords, 50):
```
This function is the one doing the slicing `grouper`.

amaltaro reviewed (changes requested):
I put a comment in the code pointing you to the slicing logic. I'd suggest to have it back in this PR to avoid having many PR adding/removing the same code, anyways this will have to be properly tested, so we better add those 10 line changes and test it altogether.

vkuznet commented:
Alan, I put back the slicer, but I really doubt it is proper solution. Instead, we should convert RequestInfo into generator, then it will not consume memory and neither being prompt to amount of records it process. The loop will be one pass across passed records where it will enhance them (if necessary) and yield back new set of records. I can do that code-refactoring later.

So far I suggest to finish this code refactoring and put it in place that we can work on top of it.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSMonitor.py:
```diff
@@ -0,0 +1,196 @@
+        """
+        return self.reqmgrAux.getTransferInfo(wname)
+
+    def getTransferIds(self, dataset):
```
This should be called `getTransferStatus` or something like that. We'll provide a transfer/request ID, and get the transfer status back (method input is wrong as well).

vkuznet replied:
the function literally returns transfer Ids, why do you need to rename it? I don't mind but I want to understand your logic.

And, if we'll provide request ID, then I need to know two things:
- how to get requestID  from WMA request name
- and, I think we should store this requestID in transfer records, otherwise we lost this info and can't use it. So far we store dataset, timstamp, transferIDs

Once you'll provide this info I'll be happy to change the name of API and its input/logic to use requestID

amaltaro replied:
This is exactly what we discussed over that long private thread.
Things from the top of my head - but better getting back there to review what I say - the PhEDEx `subscribe` API returns the request ID (which I also call transfer ID).

As we discussed about the transfer document schema, this request ID is supposed to be persisted in Couch, according to this structure:
https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154#file-transferor_format-json-L32

vkuznet replied:
Not so fast. Did you ever check what phedex subscriptions API return for given request? Please give it a try, e.g. this url (using dataset input parameter):
https://cmsweb-testbed.cern.ch/phedex/datasvc/xml/prod/subscriptions?dataset=/EG/Run2010A-WZEG-Nov4Skim_v1/RAW-RECO
will return proper subscription detials in no time (call time is 0.016 sec), while this call (using request ID)
https://cmsweb-testbed.cern.ch/phedex/datasvc/xml/prod/subscriptions?request=191805
will return no results in 45 seconds. May be I'm doing something wrong but so far usage of request ID does not seem to work with Phedex subscriptions API

amaltaro replied:
Interesting, I thought we could simply provide the request ID parameter to that API, but indeed it takes a long time to resolve it (even with a newer example/subscription).
It looks like we need to provide either the dataset or the block name, together with the request ID. In that case, we can quickly get a "valid" output, e.g.:
https://cmsweb-testbed.cern.ch/phedex/datasvc/json/prod/subscriptions?request=1857599&block=/RelValH125GGgluonfusion_14/Integ_Test-SC_LumiMask_PhEDEx_HG1908_Val_Alanv11-v11/NANOAODSIM%23*

We might have to ping Natalia on this item once again.

vkuznet replied:
so, what should we do? So far I left everything as is, i.e. I used dataset as input parameter. Until you clarify how this should be done I don't think we need to change anything.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSMonitor.py:
```diff
@@ -0,0 +1,196 @@
+            # obtain new transfer ids for given request name and dataset
+            tids = self.getTransferIds(dataset)
+            rec = {'dataset': dataset, 'dataType': dtype, 'transferIDs': tids}
+            statusRecords += rec
```
this won't work with list + dict

vkuznet replied:
ok, fixed

amaltaro commented on src/python/WMCore/MicroService/Unified/MSMonitor.py:
```diff
@@ -0,0 +1,196 @@
+                    return False
+        return True
+
+    def checkStatusRecords(self, records):
```
I'm still not sure I get the usefulness of this method.

vkuznet replied:
read the code. It does the following:
- read given records and extract dataset
- it gets new transfer IDs (transfer statuses) for a given dataset
- it save them as new record
- it calculates completion value
- it returns new transferIds (transferStatus) and completion value

amaltaro replied:
Which is wrong, see the MS Monitor wiki I provided before (point 4).
BTW, this method is apparently doing the same thing as `getStatusRecords`.

vkuznet replied:
And, I insist that we need to get complete status value somehow. I don't mind to remove this method, but you should tell me how to get complete value then?

It is not about persisting records in CouchDB (which is what point 4 is saying) it is about how to find if we completed all transfers. For that we need to take records we put into CouchDB and check them again somehow. The former task (how to get records from Couchdb) is done in `getStatusRecords` while the later task (how to obtain complete status) is done in `checkStatusRecords`. While we obtain a new transfer completion values we can get new transfer IDs (request IDs). That is why checkStatusRecrods is doing both tasks.

vkuznet replied:
Let me rephrase it. Monitor wiki says:
```
using the transfer ids available in the transfer document, make calls to PhEDEx/Rucio in order to get the status of those transfers; then calculate the transfer status completion. 
```
The `checkStatusRecords` is doing exactly this:
- take status records (transfer documents)
- it calls `getTransferIds` APi for each record which by itself
- calls PhEDEx subscriptions API
- find out (calculate) completion value

I dont' understand do you want to get rid of this method, rename it or change its logic.

amaltaro replied:
I'd rather not mix status records with transfer records. The way I see it is:
* getTransferIds: retrieve the transfer IDs for a given workflow
* checkStatusRecords (or renamed to checkTransferCompletion): uses the transfer IDs, query PhEDEx and calculate the dataset/transfer completion (can be done in multiple methods if you like).

So I see no use for `getStatusRecords`. In addition to that, `checkStatusRecords` is not supposed to generate any new transfer ID, it's supposed to receive transfer IDs, and treat them.

vkuznet replied:
ok, I did the following:
- renamed `checkStatusRecords` to `completion` method, so far it does very simple operation. I decided to have it if we need to implement more complex algorithm.
- new method takes transfer records and return completion value

amaltaro commented on src/python/WMCore/MicroService/Unified/MSMonitor.py:
```diff
@@ -0,0 +1,196 @@
+                statusRecords, completed = self.checkStatusRecords(statusRecords)
+                # if all transfers are completed
+                # move the request status staging -> staged
+                if completed == 100:
```
This is 100 right now, but soon we'll have to compare against the numbers defined in the campaign configuration (and perhaps the unified config too, I don't know yet).

vkuznet replied:
Done, via this change:
https://github.com/dmwm/WMCore/pull/9276/files#diff-4cd93292a8b855fbad20e8e899adb8b7R46

amaltaro commented on src/python/WMCore/MicroService/Unified/MSMonitor.py:
```diff
@@ -0,0 +1,196 @@
+                # if pileup transfers are completed AND
+                # some input blocks are completed,
+                # move the request status staging -> staged
+                elif self.pileupTransfersCompleted(reqName):
```
This check has likely to be done together with the input data.

vkuznet replied:
what IS the input data? please be explicit.

amaltaro replied:
point 2. here lists what "input data" can be:
https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Transferor#relying-on-reqmgr2-second-proposal

vkuznet replied:
if I read wiki properly the point 2 tells me that input data is a list of datasets (either primary, parent, or pileup datasets). But the point is that you insisted in code to use this call
```
requests = self.reqmgr2.getRequestByStatus([reqStatus], detail=False)
```
see line https://github.com/dmwm/WMCore/pull/9276/files#diff-4cd93292a8b855fbad20e8e899adb8b7R32
This call returns ONLY request names and other information. How can I obtain the input data for a given request? Do I need to make another call? Or, should I change `getRequestByStatus` details to True to get all details? In other words if you want to use input data (list of datasets from request) then we should get them somehow.

Please let me know how you'd like to resolve this.

amaltaro replied:
That's right, "input data" is a list of datasets.
Getting requests with "detail=false" returns only the request name indeed, but that's why I said we need to then retrieve the transfer record (which is named after the workflow name), that transfer document will have all the input data related information that we need (including transfer IDs).

In short,
1. we retrieve workflow names in the `staging` status
2. then we retrieve its transfer document in order to proceed with the input dataset monitoring/check (note here there is a PS., as mentioned somewhere else in this PR about bulk get)

vkuznet replied:
ok, not it is clear. I changed input parameter of pileup method and implemented basic logic according to wiki (it may not be complete though).

amaltaro commented on src/python/WMCore/MicroService/Unified/MSMonitor.py:
```diff
@@ -0,0 +1,196 @@
+        https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154
+        So far logic is based on option D of the records
+
+        .. doctest::
```
what 's this `.. doctest::` for?

vkuznet replied:
It is sphinx placeholder for code (cut-and-paste feature), see https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html?highlight=doctest#doctest-blocks

amaltaro reviewed (changes requested):
Valentin, I made further comments in the code and in general it looks Okay. I don't think we need to get a perfect logic for the MSMonitoring here because it's not really the purpose of this PR AFAICS.

Besides those comments, I think we should fetch the transfer records in bulk such that we can avoid hundreds/thousands of calls, so in the MSMonitor, you fetch all the requests in a given status, then you fetch all the transfer. Amount of data should be pretty small (I foresee <20MB).

Please do review the MSMonitor logic though:
https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Monitor

vkuznet commented:
Alan, regarding your comment:
```
Besides those comments, I think we should fetch the transfer records in bulk such that we can avoid hundreds/thousands of calls, so in the MSMonitor, you fetch all the requests in a given status, then you fetch all the transfer. Amount of data should be pretty small (I foresee <20MB).
```
which bulk operation are you referring to? Please be explicit. The code now uses `getTransferInfo` API from ReqMgrAux. This API has workflow name as an input parameter. If you want to get all records from CouchDB you need to provide appropriate API for that. Otherwise I don't see ANY way to get all records (or do bulk operation). If you're talking about another topic please provide pointer to API(s) which does this bulk operation.

amaltaro reviewed (changes requested):
Further comments in the code. If something is not clear - after you read the wiki - please let me know.
Otherwise, update the code and give it a very basic test please. If everything is fine, we might be able to request another testbed upgrade tomorrow morning.

amaltaro commented:
Valentin, about getting the transfer docs in bulk, it can be done using the same `getTransferInfo` API from ReqMgrAux module, by providing `ALL_DOCS` as the document name.

I updated this information here as well:
https://github.com/dmwm/WMCore/wiki/ReqMgr2-apis#retrieve-a-workflow-transfer-document

vkuznet commented:
ok, here is what I've done:
- use bulk call via ALL_DOCS, I'm not convinced that it will be faster and I put proper comment in a code
- I renamed checkRecordsStatus to completion method, and make later accept transfer records to find out completion value
- I renamed statusRecords into transferRecords for better code clarify
- I implemented basic pileup logic based on wiki description (further tweaking may be required)
- I left as is subscriptions call since Phedex API does not provide call via requests only, at the moment I don't see need to change anything until it is clarified
- added completion threshold , read it from configuration

I think at this moment all requests are implemented. Feel free to review again.

amaltaro commented:
Thanks, Valentin. I'll look into it tomorrow.
Just one more comment. The completion threshold does *not* come from the service configuration. It will either come from the unified config or from the campaign one.

Last but not least, can you please make a nice description in the first comment of this PR explaining what are the changes and so on? Thanks

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -1,102 +1,92 @@
-        """
-        pass
+        # update/insert requestStatues in couchdb
+        self.updateTransferInfo(requestStatuses)
```
It crashes in this line.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSTransferor.py:
```diff
@@ -1,102 +1,92 @@
         except Exception as err:  # general error
             self.logger.exception('### transferor error: %s', str(err))
+            # compose status record
+            wname = reqName  # in our case workflow name is identical to request name
```
UnboundLocalError: local variable 'reqName' referenced before assignment

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSMonitor.py:
```diff
@@ -0,0 +1,190 @@
+        wname = reqName
+        # get existing transfer IDs record
+        doc = self.getTransferInfo(wname)
+        # loop over all records and obtain new transfer IDs
```
```
if not doc:
   self.logger.error("Failed to find a transfer document for request: %s.", wname)
   return transferRecords
```

or something like that. Then we need to skip this workflow on the for loop calling this function.

vkuznet replied:
done

amaltaro commented:
Valentin, just fix those problems such that nothing crashes, and we can proceed and merge it.
I'll then work on their logic in another PR.

vkuznet commented:
Alan, I fixed issues you pointed out and it require additional code refactoring (I was need to move common methods from MSMonitoring/MSTransferor to MSCore).

vkuznet commented:
I also fixed reading threshold from unified config and updated description of this issue.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSMonitor.py:
```diff
@@ -0,0 +1,134 @@
+
+            # FIXME: completion threshold should come either from unified
+            # or campaign configuration
+            thr = self.unifiedConfig().get('completedThreshold', 100)
```
please comment this line and hard-wire it to 100 (I'll come back to it tomorrow or so), otherwise...
```
2019-08-01 16:04:03,496:ERROR:MSMonitor: +++ monit error: 'list' object has no attribute 'get'
Traceback (most recent call last):
  File "/data/srv/HG1908d/sw/slc7_amd64_gcc630/cms/reqmgr2ms/0.1.2.pre3/lib/python2.7/site-packages/WMCore/MicroService/Unified/MSMonitor.py", line 35, in execute
    thr = self.unifiedConfig().get('completedThreshold', 100)
AttributeError: 'list' object has no attribute 'get'
```

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSMonitor.py:
```diff
@@ -0,0 +1,134 @@
+    def pileupTransfersCompleted(self, transferRecords):
+        "Check if pileup transfers are completed for given transfer records"
+        # loop over all records and obtain new transfer IDs
+        thr = self.msConfig.get('completedThreshold', 100)
```
same comment here, please hard-wire it to 100. It's anyways wrong, it should use unified config, not the microservice one.

vkuznet replied:
done

amaltaro reviewed (changes requested):
Valentin, there are two more lines to fix. Please just hard-code those things because I want to have this PR in asap. Such that we can start working heavily on the logic.

vkuznet commented:
Alan, I put thr = 100 in a code and put in comments that we should fix it and provide example.
Feel free to proceed.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSCore.py:
```diff
@@ -0,0 +1,130 @@
+        Fetches the unified configuration
+        :return: unified configuration content
+        """
+        return self.reqmgrAux.getUnifiedConfig(docName="config")
```
```suggestion
        res = self.reqmgrAux.getUnifiedConfig(docName="config")
        if res:
            return res[0]
        else:
            return {}
```

vkuznet replied:
I fixed it in more conservative way, I added check to res to be list.

In general, return types of WMCore APIs is somewhat fragile and inconsistent, but it is a different story.

vkuznet commented:
Alan, I submitted the fix for unifiedConfig.

amaltaro commented:
Thanks, it seems to be working fine now. Once we hear back from Jenkins, I'll merge it and put in a new tag for Lina for testbed tomorrow.
